File: 剑指offer/4.二维数组中的查找/4.二维数组中的查找.go

/*
行和列都是单调递增
x,y就相当于2个指针
从右上角开始
    - 如果从左上角开始,无法排除当前所在的行和列
遇到大于的值，就 x-=1
遇到小于的值，就 y+=1
*/

func findNumberIn2DArray(matrix [][]int, target int) bool {
    m := len(matrix)
    if m == 0{
        return false
    }
    n := len(matrix[0])
    if n == 0{
        return false
    }
    x := n - 1
    y := 0
    for y < m && x > -1{
        if matrix[y][x] < target{
            y += 1
        }else if matrix[y][x] > target{
            x -= 1
        }else{
            return true
        }
    }
    return false
}

File: 剑指offer/4.二维数组中的查找/4.二维数组中的查找.py
from typing import List


class Solution:
    def findNumberIn2DArray(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix or not target:
            return False
        m, n = len(matrix), len(matrix[0])
        row, col = 0, n - 1
        while row < m and col > -1:
            if matrix[row][col] < target:
                row += 1
            elif matrix[row][col] > target:
                col -= 1
            else:
                return True
        return False
